### DISCUSSION TITLE
General discussion about safety in v's language design

### DISCUSSION THREAD

**Discussion by nova-nowiz** [vlang/v#7457] [Ideas]
To begin with, I'd like to say that I'm very surprised by the quality of this language and that it is in my opinion almost perfect.
The thing that makes me say that v is not the perfect language is in regards to safety.

## Background
A little bit of background first. I have coded multiple projects in [Elm](elm-lang.org). Elm is a web front-end language that has the following mantra: "no runtime exceptions" and this holds true for every program you write in Elm. If it compiles, it just works.

This is something I'm pursuing in application programming languages and I did not find any language other than elm that has this premise. V could be a really really good candidate for such a thing.

## Panics
First of all, panics. I strongly agree with the points given by @dumblob in #2030. panics are less powerful and controlled than errors. This is in my opinion the next 1 million dollars mistake. It is very tempting to implement an exception style approach as it is at first way easier, but it means more work for the user overall and the code is far less expressive than without them.
Elm is a proof of the fact that one can build software without panics and only with Result/Option return values.
They must be treated by the main program and thus the program never produce any exceptions whatsoever.
This eliminates a lot of headaches for the programmer as he doesn't have to treat all sorts of exceptions implicitly thrown by packages that he includes in his work.
He doesn't have to look at documentation, only the return type of the function he calls is important.

## Mutation, Concurrency and Channels
My second point would be on mutating elements on threads. This should not be allowed is safe code IMHO.
Channels are very easy to work with and provide the same functionality: passing data from one thread to another, but without any of the risks of standard mutable data shared between thread and also without all the mutex complexity and potential problems it may generate.

## Mutable arguments in functions
Thirdly, functions with `mut` arguments seem to not be a good idea except in methods to modify the instance. and this is something that you state as well on the documentation:

> Use `user.register()` or `user = register(user)` instead of `register(mut user)`.

> It is preferable to return values instead of modifying arguments. Modifying arguments should only be done in performance-critical parts of your application to reduce allocations and copying.

Making `register(mut user)` an unsafe expression seems like a good idea as the documentation implies in the first place that it is such an expression (not recommended, used only for performance-critical parts), and I agree with it if it was the intention.

## Conclusion
A possible implementation could be at first a `-safer` flag passed to the compiler for additional check if these particular features are not agreed upon by everyone.
I however think that these changes would greatly improve this beautiful language and would benefit applications made in v as the language would be more resilient.

**Comment by nova-nowiz** [comment]
Here are a few people in the rust community stating that the `panic!` macro was a bad idea and that it defeated the purpose of using Option/Result types in the first place:
[non panicking rust](https://www.reddit.com/r/rust/comments/9mj5xc/nonpanicking_rust/) 
[#rust2018: no need to panic](https://www.reddit.com/r/rust/comments/7tcyh1/rust2018_no_need_to_panic/)
[no panic](https://internals.rust-lang.org/t/no-panic/1356)

**Comment by medvednikov** [comment]
@Narice we only use panics for index out of range errors in arrays.

Can't think of anything else.

Everything is handled by the builtin and simple option/result.

Well the user can always panic of course: `foo() or { panic(err) }`

**Comment by nova-nowiz** [comment]
For index out of range it would be possible to use results as well. It would avoid crashes and users would implement exception handling as well when using the function which make programs more robust IMHO.
I know that it is less direct but it is certainly safer.
I also understand that for some people this is too safe at the detriment of usability (even tho I think that with the or syntax it would be pretty slick), this is why I was proposing a `-safer` flag to put this as an additional restriction if wanted instead of a restriction for everyone. Everything that can panic is put in an unsafe block and then it compiles.
I don't know if this was really clear but this means that either when accessing an array element, it provides a result (meaning modification of the language) or the call for accessing an array element is put in an unsafe block by the user.

**Reply by danieldaeschle** [reply]
I agree with you. But i also think `arr[0] or { MyStruct{} }` doing all the time is annoying.

**Reply by nova-nowiz** [reply]
Yeah that's the thing...
Crystal uses unions of types for solving this issue and lobster uses flow typing (dynamic typing at compile time? Strange stuff).
Basically, nil exists as a type, but it's not possible to dereference it. The error is thrown at compile time for both languages.
These are solutions that might not be applicable to v but i'm pretty sure we could find something to avoid that sort of syntax.

**Reply by danieldaeschle** [reply]
Can you make an example?

**Reply by nova-nowiz** [reply]
[Here](https://crystal-lang.org/reference/getting_started/cli.html#request-for-user-input)  is an example from crystal
[Here](http://aardappel.github.io/lobster/type_checker.html) is the type checker explanation of lobster. The part called "The trouble with nil" is the one you should read and explains the concept of lobster's type system (it's quite similar in spirit to crystal but very different in practice)

**Reply by danieldaeschle** [reply]
yeah, but we don't have nil.

**Reply by nova-nowiz** [reply]
Yes, that's what I was trying to state by:
> These are solutions that might not be applicable to v but i'm pretty sure we could find something to avoid that sort of syntax.

If you wanted an example of what could be done instead of examples of the two languages I cited, I don't have any, I'll try to find something if possible ;)

**Reply by danieldaeschle** [reply]
thank you anyway :D

**Comment by medvednikov** [comment]
we are actually introducing an optional `m[key] or {}` to maps, will do the same for arrays

and I see no reason not to introduce `-safer` to force this usage

but definitely not by default

**Comment by nova-nowiz** [comment]
Nice thank you so much! This is really cool!

**Comment by nova-nowiz** [comment]
Btw, what do you think of the other parts?
Could it be also added to the safer flag?

**Comment by ntrel** [comment]
> For index out of range it would be possible to use results as well.

This is too inefficient. I think Alex means `or` will be supported but not required.

> m[key] or {} to maps, will do the same for arrays

What if the element type is itself an optional? (That should be supported e.g. for optional references to avoid null).

**Reply by medvednikov** [reply]
Yes, I wrote "optional or {}"

> What if the element type is itself an optional?

Not sure yet.

**Reply by dumblob** [reply]
>What if the element type is itself an optional?

Hm, I thought *V* isn't designed for reified optionals. I'm strongly convinced reified optionals (i.e. having an optional as an element type) are not a good idea.

@ntrel what would optional references be good for? A null reference is just a plain reference. Only certain APIs try to make it "magical" and say it doesn't belong to the the same class of references as other references do, but that's plain nonsense. There is actually no distinction - it supports all operations (i.e. is fully compatible with) any other reference does. Why would anyone want to distinguish a null reference on type level is still a huge uknown for me.

**Reply by medvednikov** [reply]
@dumblob my initial thoughts are the same. `?&Foo` should simply not be allowed, there's no need in it `?Foo` is enough.

**Reply by danieldaeschle** [reply]
we use `?&Foo` in the compiler (scopes).
Isn't it for performance?

**Reply by dumblob** [reply]
@medvednikov my thought was slightly different and I might have expressed it wrong. I meant that **an element type itself is not allowed to be an optional**. I.e. one can't store an optional into a map under a given key and at the same time can't withdraw any such optional from the map if given the right key.

----
As a side note, I think `?&Foo` is actually needed. E.g. due to performance as @danieldaeschle pointed out or for propagation of `error()`:

There is though one use case I can think of where `?&Foo` would be required - namely for propagation of an `error()` from a subsequent call:

```
fn f() ?&Foo {
  return_optional()?  // this is the sole reason to make
                      // the return type of `f` an optional
  return &Foo{}
}
```

**Comment by ntrel** [comment]
A mutable parameter is no less memory-safe than a mutable receiver. Mutable parameters are important for efficiency. V should be able to replace C.

Panic is important because an API might have been designed without optionals, then later an error case needs to be added. It's not practical to require all users of the API to quickly update their code to optionals.

**Reply by dumblob** [reply]
>Panic is important because an API might have been designed without optionals, then later an error case needs to be added. It's not practical to require all users of the API to quickly update their code to optionals.

That's a very important argument and I agree. I'd though expect the compiler to tell me where everywhere there are panics (including stdlib and V's internals) if I asked for that (e.g. using the `-safer` flag).

**Reply by nova-nowiz** [reply]
> A mutable parameter is no less memory-safe than a mutable receiver. Mutable parameters are important for efficiency. V should be able to replace C.

If you are talking about using them in a single thread, well my bad, you are right. Mutable parameters are unpure but not unsafe in a single thread and I was wrong x).
However, if the discussion is on multiple thread, mutable parameters are very unsafe and requires additional measures like mutex which are not necessary if you use channels. CSP is a better option in my opinion than having plain shared memory. Shared memory outside of channels should be marked as unsafe. This allows users to use it for performance critical parts of their app, but it is explicitly marked as unsafe so that they know that this code can be unsafe and that all the other code is not the reason of the bug which drastically reduced the mental charge for the user while debugging.

**Reply by nova-nowiz** [reply]
> Panic is important because an API might have been designed without optionals, then later an error case needs to be added. It's not practical to require all users of the API to quickly update their code to optionals.

I might have not understood correctly this part so just tell me if it is not what you meant.
It seems that you are saying that:
- some APIs are design without results
- APIs upgrading from optionals to results force users to also upgrade
Yes, this is the case, but it could be a language endeavor instead of an API endeavor.

@medvednikov what tools are we allowed to use to pass from one state of the language to another? I know using vfmt is allowed but is it possible to use the compiler if the process is painless?

If we use the compiler, it is very painless as you only need to follow the compiler (as in Elm) and it will just tell you what to change and where.
I don't know what vfmt can and can't do but if it is possible to get info from the compiler, that might make it easy (but that's only an educated guess)

**Comment by nova-nowiz** [comment]
After some reflection, I disagree with myself on my third point (mutable parameters in functions).
They are unpure, yes, but not dangerous if in a single thread.
So the real culprite here is using them in other threads (my second point).
